qiime2/plugin/model/tests/test_directory_format.py

```diff
@@ -0,0 +1,104 @@
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+
+import unittest
+
+import pkg_resources
+from qiime2.plugin import model
+
+from qiime2.core.testing.format import IntSequenceFormat
+from qiime2.core.exceptions import ValidationError
+
+
+# Define dummy plugin formats to test with
+
+class AllRequiredDirFmt(model.DirectoryFormat):
+    file1 = model.File(r'test_text1.txt', format=IntSequenceFormat,
+                       optional=False)
+    file2 = model.File(r'test_text2.txt', format=IntSequenceFormat,
+                       optional=False)
+    file3 = model.File(r'test_text3.txt', format=IntSequenceFormat,
+                       optional=False)
+
+
+class AllRequiredDefaultDirFmt(model.DirectoryFormat):
+    file1 = model.File(r'test_text1.txt', format=IntSequenceFormat)
+    file2 = model.File(r'test_text2.txt', format=IntSequenceFormat)
+    file3 = model.File(r'test_text3.txt', format=IntSequenceFormat)
+
+
+class OptionalDirFmt(model.DirectoryFormat):
+    file1 = model.File(r'test_text1.txt', format=IntSequenceFormat,
+                       optional=False)
+    file2 = model.File(r'test_text2.txt', format=IntSequenceFormat,
+                       optional=False)
+    file3 = model.File(r'test_text3.txt', format=IntSequenceFormat,
+                       optional=True)
+
+
+class TestDirectoryFormat(unittest.TestCase):
+    package = 'qiime2.plugin.model.tests'
+
+    def get_data_path(self, filename):
+        """Convenience method for getting a data asset while testing.
+
+        Test data stored in the ``data/`` dir local to the running test
+        can be accessed via this method.
+
+        Parameters
+        ----------
+        filename : str
+            The name of the file to look up.
+
+        Returns
+        -------
+        filepath : str
+            The materialized filepath to the requested test data.
+
+        """
+
+        return pkg_resources.resource_filename(self.package,
+                                               'data/%s' % filename)
+
+    def test_fails_missing_required(self):
+        files_dir_fp = self.get_data_path('test_text_files/')
+
+        with self.assertRaisesRegex(
+            ValidationError, "Missing one or more files for"
+                             " AllRequiredDirFmt"):
+
+            format_object = AllRequiredDirFmt(
+                                files_dir_fp,
+                                mode='r',
+                                )
+
+            format_object.validate()
+
+    def test_fails_missing_with_optional_default(self):
+        files_dir_fp = self.get_data_path('test_text_files/')
+
+        with self.assertRaisesRegex(ValidationError,
+                                    "Missing one or more files for "
+                                    "AllRequiredDefaultDirFmt: "
+                                    "\'test_text3.txt\'"):
```

Comment:
This assertion is brittle to the sort order of the internal logic. I would make it similar to your above test and omit the specific filename missing (or make the directory format simpler with only a single property instead of 3).